### PR TITLE
Create API resources for DeploymentTarget/Claim/Class

### PR DESCRIPTION
**Pull request by chetan-rns**
- Use operator-sdk to create CRs for DeploymentTarget, DeploymentTargetClaim, and DeploymentTarget.
- Add fields to CRs based on https://redhat-appstudio.github.io/book/ADR/0008-environment-provisioning.html

Jira: https://issues.redhat.com/browse/GITOPSRVCE-386

### CODE REVIEW

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great, thanks @chetan-rns!

Just a couple of tiny nits: We can remove the `EDIT THIS FILE! THIS IS SCAFFOLDING FOR YOU TO OWN!` line, and it couldn't hurt to update the copyright to 2023. 

Then it's good to go.

**Comment by chetan-rns** [comment]
@jgwest I've removed the comment and updated the year. PTAL

File: api/v1alpha1/deploymenttargetclaim_types.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// DeploymentTargetClaimSpec defines the desired state of DeploymentTargetClaim
+type DeploymentTargetClaimSpec struct {
+	DeploymentTargetClassName DeploymentTargetClassName `json:"deploymentTargetClassName"`
+}
```
Comment (jgwest) [inline api/v1alpha1/deploymenttargetclaim_types.go:28]: I believe DeploymentTargetClaim should also have a `targetName` field. This field isn't referenced in any of the examples, but it is references in the flowcharts/images from Miro. I talk a bit more about this in [this doc](https://hackmd.io/@_Ujs_9QpQgyQYgLo2maa2Q/ByXcz-20o)

File: api/v1alpha1/deploymenttargetclaim_types.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+type DeploymentTargetClassName string
+
+const (
+	DTCTName_IsolationLevelNamespace DeploymentTargetClassName = "isolation-level-namespace"
```
Comment (jgwest) [inline api/v1alpha1/deploymenttargetclaim_types.go:33]: I believe this constant doesn't need to be defined in this package, and I don't think there is anything special about the `isolation-level-namespace` string constant.

`isolation-level-namespace` just happens to be the name of the `DeploymentTargetClass` they use in their example in ADR-8 doc. In practice, one could use any name for the DTClass, as long as it has this field `provisioner: appstudio.redhat.com/devsandbox`

File: PROJECT [hunk 3]
```diff
@@ -82,4 +82,28 @@ resources:
   kind: Environment
   path: github.com/redhat-appstudio/application-api/api/v1alpha1
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
```
Comment (jgwest) [inline PROJECT:87]: `DeploymentTargetClass` should be cluster-scoped, rather than namespace-scoped, it looks like.

**Review by jgwest — CHANGES_REQUESTED** [review]

File: api/v1alpha1/deploymenttarget_types.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// DeploymentTargetSpec defines the desired state of DeploymentTarget
+type DeploymentTargetSpec struct {
+	DeploymentTargetClassName DeploymentTargetClassName `json:"deploymentTargetClassName"`
+
+	KubernetesClusterCredentials DeploymentTargetKubernetesClusterCredentials `json:"kubernetesCredentials"`
+
+	ClaimRef string `json:"claimRef"`
```
Comment (jgwest) [inline api/v1alpha1/deploymenttarget_types.go:31]: Can the `claimRef` be optional in some cases, eg it should have the `omitEmpty` tag?

File: api/v1alpha1/deploymenttargetclaim_types.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// DeploymentTargetClaimSpec defines the desired state of DeploymentTargetClaim
+type DeploymentTargetClaimSpec struct {
+	DeploymentTargetClassName DeploymentTargetClassName `json:"deploymentTargetClassName"`
+	TargetName                string                    `json:"targetName"`
```
Comment (jgwest) [inline api/v1alpha1/deploymenttargetclaim_types.go:28]: I think the `targetName` field can be optional in some cases, eg it should have the `omitEmpty` tag (feel free to correct me if I am wrong)

**Review by jgwest — CHANGES_REQUESTED** [review]
A couple more things I noticed, then good to go.

File: api/v1alpha1/deploymenttarget_types.go (hunk 4 above)
Comment (chetan-rns) [reply api/v1alpha1/deploymenttarget_types.go:31]: Good catch. It could be optional

**Review by jgwest — APPROVED** [review]
Excellent, thanks @chetan-rns!

**Pull request merged**